### PR TITLE
[Dist Pass] Amp Pass

### PR DESCRIPTION
**Pull request by JZ-LIANG**
### PR types
New features

### PR changes
Others

### Describe
* amp pass supports for auto parallel (support dist op and dist context)
* different from fleet.meta_optimizer.amp which uses nested callback for modification of forward/backward/update. this amp pass modify the full program(forward/backward/update) in once, which increases the independency and maintainability.
* the performance is the same as fleet.meta_optimizer.amp

### CODE REVIEW

File: python/paddle/distributed/auto_parallel/parallelizer.py [hunk 1]
```diff
@@ -185,10 +188,9 @@ def _get_dist_program(self, rank, dist_context=None, relaunch_phase=False):
             self._parameter_list, self._no_grad_set, self._callbacks)
 
         # serial forward pass
-        self._apply_serial_pass(completed_main_program, serial_startup_program)
-
+        self._apply_serial_pass(completed_main_program, serial_startup_program,
```
Comment (aoyulong) [inline python/paddle/distributed/auto_parallel/parallelizer.py]: Please change `_apply_serial_pass()` to `_apply_pre_passes()` and change `_apply_post_optimization_passed` to `_apply_post_passes`.

File: python/paddle/distributed/auto_parallel/operators/dist_check_finite_and_unscale.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+from paddle.fluid.framework import in_dygraph_mode
+from paddle.fluid.data_feeder import check_variable_and_dtype, check_dtype
+from paddle.distributed.fleet.meta_optimizers.common import OpRole, OP_ROLE_KEY, OP_ROLE_VAR_KEY
+from ..utils import set_var_dist_attr
+from ..utils import set_dist_op_desc_original_id
+from ..process_group import new_process_group
+from ..dist_attribute import OperatorDistributedAttribute
+from paddle.distributed.auto_parallel.process_group import get_world_process_groups
+
+global_process_mesh = get_world_process_groups().ranks
```
Comment (aoyulong) [inline python/paddle/distributed/auto_parallel/operators/dist_check_finite_and_unscale.py]: Please change `get_world_process_groups()` to `get_world_process_group()`.

File: python/paddle/distributed/passes/auto_parallel_amp.py [hunk 3]
```diff
@@ -0,0 +1,715 @@
...
+        for g in grads:
+            g_dist_attr = self.dist_context.get_tensor_dist_attr_for_program(g)
+            assert g_dist_attr is not None
+            new_op_dist_attr.set_input_dims_mapping(g.name,
+                                                    g_dist_attr.dims_mapping)
+            new_op_dist_attr.set_output_dims_mapping(g.name,
+                                                     g_dist_attr.dims_mapping)
+        self.dist_context.set_op_dist_attr_for_program(new_op, new_op_dist_attr)
+
+        main_block._sync_with_cpp()
```
Comment (aoyulong) [inline python/paddle/distributed/passes/auto_parallel_amp.py:715]: Please add the necessary comments, especially for the methods without underscores.

**Review by aoyulong — DISMISSED** [review]

File: python/paddle/distributed/auto_parallel/parallelizer.py (hunk 1 above)
Comment (JZ-LIANG) [reply python/paddle/distributed/auto_parallel/parallelizer.py]: fixed~

File: python/paddle/distributed/auto_parallel/operators/dist_check_finite_and_unscale.py (hunk 2 above)
Comment (JZ-LIANG) [reply python/paddle/distributed/auto_parallel/operators/dist_check_finite_and_unscale.py]: fixed, though it is not my fault~  lol

File: python/paddle/distributed/passes/auto_parallel_amp.py (hunk 3 above)
Comment (JZ-LIANG) [reply python/paddle/distributed/passes/auto_parallel_amp.py:715]: done~

**Review by sneaxiy — APPROVED** [review]
LGTM

**Pull request merged**